docs/Dialects/FIRRTL/FIRRTLAnnotations.md

```diff
@@ -1495,6 +1495,14 @@ Example:
 }
 ```
 
+## Discardable Annotaions
+
+By default, annotations in the FIRRTL dialect prevent dead code elimination (DCE).
```

Comment:
Nit: This is specific to dead-code elim in particular.  Does this literally only impact whether a value is considered "live" (if the annotation pointing to it is also live)? That does seem more straightforward re:scope.

"Discardable" suggests that it could be dropped, perhaps for no reason (?), but certainly by various optimizations such as canonicalizers.  Is this right?

Are there requirements to preserve this /if/ the target is preserved in some sense (something something notions of equality/identity)?  For example, could we CSE a value with this annotation, or discard it during a transform if inconvenient to preserve?  Where in the spectrum between "if this exists (asterisks on what precisely that means), the annotation will be on it I promise" vs "this is like a namehint, we do what we can but ultimately who can say and if you need this for correctness do it a different way"?

Targeting DCE specifically mostly avoids these questions, although it seems like defining the semantics this way is very specific to our implementation and notion of "DCE".

Do annotations block, e.g., IMCP, generally (if not, do discardable annotations block IMCP?)? Presumably if we can DCE this we can also constant-prop through it.

Reply:
> "Discardable" suggests that it could be dropped, perhaps for no reason (?), but certainly by various optimizations such as canonicalizers. Is this right?

Yes, my intention was canonicalizer or folder can also drop the annotation if there is no user of the annotation target.\


> For example, could we CSE a value with this annotation, or discard it during a transform if inconvenient to preserve? 

I want to prohibit these operations.

> Where in the spectrum between "if this exists (asterisks on what precisely that means), the annotation will be on it I promise" vs "this is like a namehint, we do what we can but ultimately who can say and if you need this for correctness do it a different way"?

The former semantics is exactly what I want to represent.

> Do annotations block, e.g., IMCP, generally (if not, do discardable annotations block IMCP?)? Presumably if we can DCE this we can also constant-prop through it.

Annotations usually don't block IMCP. Only dontTouch annotation blocks the IMCP.